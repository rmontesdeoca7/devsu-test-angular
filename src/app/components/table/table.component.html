<div class="main">
  <table class="table">
    <thead>
      <tr>
        <th>Logo</th>
        <th>Nombre del producto</th>
        <th>Descripción 
          <div class="tooltip">
            <img src="assets/img/tooltip.svg" alt="tooltip">
            <span class="tooltiptext">Descripción</span>
          </div>
        </th>
        <th>Fecha de liberación 
          <div class="tooltip">
            <img src="assets/img/tooltip.svg" alt="tooltip">
            <span class="tooltiptext">Fecha de liberación</span>
          </div>
        </th>
        <th>Fecha de reestructuración 
          <div class="tooltip">
            <img src="assets/img/tooltip.svg" alt="tooltip">
            <span class="tooltiptext">Fecha de reestructuración</span>
          </div>
        </th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr *ngFor="let product of products">
        <td><img height="20" [src]="product.logo" [alt]="product.name" /> </td>
        <td>{{ product.name }}</td>
        <td>{{ product.description }}</td>
        <td>{{ product.date_release  | date: 'dd/MM/yyyy' }}</td>
        <td>{{ product.date_revision | date: 'dd/MM/yyyy' }}</td>
        <td>...</td>
      </tr>

    </tbody>
  </table>

  <footer class="footer">
    <span class="result">
      {{ productsLength }} Resultados
    </span>
    <div class="selector">
      <select >
        <option value="5">5</option>
      </select>
    </div>
  </footer>
</div>
