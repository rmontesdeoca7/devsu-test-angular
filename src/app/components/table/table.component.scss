.main {
  background-color: #fff;
  border-radius: 8px;
  padding: 1rem;

  .footer {
    display: flex;
    justify-content: space-between;
  }
}

.table {
  width: 100%;
  border-collapse: collapse;


  thead {
    background-color: #f9fafc;

    th {
      text-align: left;
      font-weight: bold;
    }
  }
  tbody td {
    padding: .5rem 0;
  }

  tr {
    height: 45px;
    border-bottom: 2px solid #f7f7f7;
  }
}