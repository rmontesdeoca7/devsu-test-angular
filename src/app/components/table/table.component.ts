import { Component, Input, OnInit, inject } from '@angular/core';
import { ApiService } from '../../services/api.service';
import { Product } from 'src/app/interfaces/products-response.interface';

@Component({
  selector: 'app-table',
  templateUrl: './table.component.html',
  styleUrls: ['./table.component.scss']
})
export class TableComponent implements OnInit {
  private apiService = inject( ApiService );
  public products:Product[] = [];
  public productsAll:Product[] = [];

  ngOnInit(){
    this.getProducts();
    this.getSearchTerm();
  }

  get productsLength () {
    return this.products.length;
  }

  getSearchTerm() {
    this.apiService.dispachSearchTerm
    .subscribe( term => {
     this.products = this.productsAll
      .filter( 
        product => 
          product.name.toLowerCase().includes(term) || 
          product.description.toLowerCase().includes(term)  
      )
    })
  }

  getProducts() {
    this.apiService.getProducts()
    .subscribe({
      next: (products) => {
        this.productsAll = products;
        this.products = products;
      },
      error: (message) => console.error({ message })
    });
  }

}
