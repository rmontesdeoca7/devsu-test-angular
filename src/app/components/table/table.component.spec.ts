import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing'
import { TableComponent } from './table.component';
import { ApiService } from '../../services/api.service';
import { ToastrService } from 'ngx-toastr';
import { Product } from '../../interfaces/products-response.interface';

describe('TableComponent', () => {
  let component: TableComponent;
  let fixture: ComponentFixture<TableComponent>;
  let toastrService: ToastrService;
  let compiled: HTMLElement;
  let service: ApiService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TableComponent],
      imports: [
        HttpClientTestingModule
      ],
      providers: [
        ApiService,
        { provide: ToastrService, useValue: toastrService }
      ]
    });
    fixture = TestBed.createComponent(TableComponent);
    component = fixture.componentInstance;
    service = TestBed.inject( ApiService );

    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should be match with the snapshop', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  xit('should call getProducts()', () => {
    const spy = jest.spyOn(service, 'getProducts').mockImplementation();
    component.productsAll = [{
      id: 'tj-q1',
      name: 'Tarjetas',
      description: 'Tarjetas de crédito',
      logo: 'img.jpg',
      date_release: '2023-07-26',
      date_revision: '2024-07-26'
    }]
    component.getProducts();
    expect(spy).toBeCalled();
  });

  it('should call onChangeSelect()', () => {
    const value = "5";
    component.productsAll = [{
      id: 'tj-q1',
      name: 'Tarjetas',
      description: 'Tarjetas de crédito',
      logo: 'img.jpg',
      date_release: '2023-07-26',
      date_revision: '2024-07-26'
    }]
    component.onChangeSelect(value);
    expect(component.products.length).toBeLessThanOrEqual(+value);
  });
});
