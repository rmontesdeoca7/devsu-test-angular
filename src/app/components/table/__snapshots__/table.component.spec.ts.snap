// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent should be match with the snapshop 1`] = `
"<div class="main"><!--bindings={
  "ng-reflect-ng-if": "false"
}--><footer class="footer"><span class="result"> 0 Resultados </span><div class="selector"><select><option value="5">5</option><option value="10">10</option><option value="20">20</option></select></div></footer></div>"
`;
